plutus-core/testlib/PlutusCore/Generators/PIR/ShrinkTypes.hs

```diff
@@ -0,0 +1,252 @@
+-- type of that kind. Note: this function requires care and attention to not get
+-- a shrinking loop. If you think you need to mess with this function:
+-- 0. First, you *must* read the note Note [Avoiding Shrinking Loops]
+-- 1. You're probably wrong, think again and
+-- 2. If you're sure you're not wrong you need to be very careful and
+--    test the shrinking to make sure you don't get in a loop.
+-- 3. Finally, you *must* read the note Note [Avoiding Shrinking Loops]
+minimalType :: Kind () -> Type TyName DefaultUni ()
+minimalType = go . argsKind where
+  go = \case
+    []               -> unit
+    [Type{}]         -> list
+    [Type{}, Type{}] -> pair
+    k:ks             -> TyLam () (TyName $ Name "_" (toEnum 0)) k $ go ks
+
+  unit = mkTyBuiltin @_ @() ()
+  list = mkTyBuiltin @_ @[] ()
+  pair = mkTyBuiltin @_ @(,) ()
+
+-- | Take a type in a context and a new target kind
+--   Precondition: new kind is smaller or equal to old kind of the type.
+--   Complies with the implicit shrinking order.
+--   TODO (later): also allow changing which context it's valid in
+fixKind :: HasCallStack
+        => TypeCtx
+        -> Type TyName DefaultUni ()
+        -> Kind ()
+        -> Type TyName DefaultUni ()
+fixKind ctx ty k
+  -- Nothing to do if we already have the right kind
+  | unsafeInferKind ctx ty == k = ty
+  | not $ k `leKind` unsafeInferKind ctx ty = error $ concat
+      [ "Internal error. New kind: "
+      , display k
+      , "\nis not smaller than the old one: "
+      , display $ unsafeInferKind ctx ty
+      ]
+  | otherwise = case ty of
+    -- Switch a variable out for a different variable of the right kind
+    TyVar _ _ -> case [y | (y, k') <- Map.toList ctx, k == k'] of
+        y : _ -> TyVar () y
+        _     -> minimalType k
+    -- Try to fix application by fixing the function
+    TyApp _ a b -> TyApp () (fixKind ctx a $ KindArrow () (unsafeInferKind ctx b) k) b
+    TyLam _ x kx b ->
+      case k of
+        -- Fix lambdas to * by substituting a minimal type for the argument
+        -- and fixing the body.
+        Type{} -> fixKind ctx (typeSubstClosedType x (minimalType kx) b) k
+        -- Fix functions by either keeping the argument around (if we can) or getting
+        -- rid of the argument (by turning its use-sites into minimal types) and introducing
+        -- a new argument.
+        KindArrow _ ka kb
+          | ka == kx              -> TyLam () x kx $ fixKind (Map.insert x kx ctx) b kb
+          | not $ kb `leKind` kb' -> error "fixKind"
+          | otherwise             -> TyLam () x ka $ fixKind ctx' b' kb
+            where
+              ctx' = Map.insert x ka ctx
+              b'   = typeSubstClosedType x (minimalType kx) b
+              kb'  = unsafeInferKind ctx' b'
+    -- Ill-kinded builtins just go to minimal types
+    TyBuiltin{} -> minimalType k
+    -- Unreachable, because there's no kind that is smaller than @Type ()@.
```

Comment:
I was confused by that too. If the kind is the same, then we always return the given type back without any adjustments.

Reply:
Ah okay, then it would be good to explain the reasoning here a bit more: "Unreachable, because the target kind must be less than or equal to the original kind. We handled the case where they are equal at the beginning of the function, so at this point the target kind must be strictly less than the original kind, but for these types we know the original kind is Type, and there is no kind smaller than that."

Reply:
Great suggestion, thanks.